File: Halo/Halo_make.m
function Ret = Halo_make(Img, CasterLevel)
Ret = [];
Oracle = round(regionprops(imbinarize(Img),'centroid').Centroid);
block = imbinarize(Img);
Bedge = edge(Img,'canny');
% Bedge = imresize(Bedge, 10);
figure; imshow(Bedge)
% BW = edge(I,'sobel');

%% Find the Oracle
Img(Oracle(1),Oracle(2)) = 0;
[edge_x,edge_y] = find(Bedge==1);
O_x = edge_x-Oracle(1);
O_y = edge_y-Oracle(2);
O_d = sqrt(O_y.*O_y+O_x.*O_x);

%% cast神圣新星：128道激光
% CasterLevel = 32;
Thita_Index = [0:CasterLevel-1]'*2*pi/CasterLevel;
temp1 = cos(Thita_Index);   % degree: 128
temp2 = O_y./O_d;           % degree: 93

% 首先区分y的正负性
[Y_posIndex, Y_pos] = find(O_y>=0);
[Y_negIndex, Y_neg] = find(O_y<0);
[temp_pos, sortposIndex] = sortrows(temp2(Y_posIndex),-1);
[temp_neg, sortnegIndex] = sortrows(temp2(Y_negIndex),-1);

%% 算法思路
% 先将图形放大（插值平滑曲线），再寻找最外层边缘
% 按照这个算法，小图是找不到最外层边缘的
% 此算法经过类人认证！！！

% 点积平方/模长
% for 128条Aegis Nova射线，遍历有序cos序列，寻找相交点

jj = 1;
for ii = 1:(CasterLevel/2)
    Ret(ii,2) = inf;
    % find(角度接近&&y>=0)
    for jj = 1:size(temp_pos,1)
        if(abs(temp1(ii)-temp_pos(jj)) < Ret(ii,2))
            Ret(ii,1) = O_d(sortposIndex(jj));
            Ret(ii,2) = abs(temp1(ii)-temp_pos(jj));
            Ret(ii,3) = sortposIndex(jj);
        end
    end
    % O(n^2)时间复杂度，需要后续优化为O(n)
end

jj = 1;
for ii = (CasterLevel/2+1):CasterLevel
    Ret(ii,2) = inf;
    % find(角度接近&&y>=0)
    for jj = 1:size(temp_neg,1)
        if(abs(temp1(ii)-temp_neg(jj)) < Ret(ii,2))
            Ret(ii,1) = O_d(sortnegIndex(jj));
            Ret(ii,2) = abs(temp1(ii)-temp_neg(jj));
            Ret(ii,3) = sortnegIndex(jj);
        end
    end
    % O(n^2)时间复杂度，需要后续优化为O(n)
end

end

File: Halo/Halo_Python/Halo_make.py
import cv2 as cv
import numpy as np

def Halo_make(Img, CasterLevel):
    xgrad = cv.Sobel(Img, cv.CV_16SC1, 1, 0)
    ygrad = cv.Sobel(Img, cv.CV_16SC1, 0, 1)
    Bedge = cv.Canny(xgrad, ygrad, 50, 150)


    Oracle = regionprops(imbinarize(Img),'centroid').Centroid
    block = imbinarize(Img)
    Bedge = edge(Img,'canny')
    # Bedge = imresize(Bedge, 10)
    # figure; imshow(Bedge)
    # BW = edge(I,'sobel')

    ## Find the Oracle
    [edge_x,edge_y] = find(Bedge==1)
    O_x = edge_x-Oracle(1)
    O_y = edge_y-Oracle(2)
    O_d = sqrt(O_y.*O_y+O_x.*O_x)

    ## cast神圣新星：128道激光
    Thita_Index = (0:CasterLevel-1).T*2*pi/CasterLevel
    for ii = 1:CasterLevel
        Falcon = sin(Thita_Index(ii)).*O_x + cos(Thita_Index(ii)).*O_y
        Falcon = Falcon .* (Falcon./O_d).^256
        
        [MaxFalcon,index] = max(Falcon)
        if MaxFalcon > 0.01
            Ret(ii,1) = O_d(index)
            Ret(ii,2) = 0
            Ret(ii,3) = index
        else
            Ret(ii,1) = 0.5
            Ret(ii,2) = 0
            Ret(ii,3) = index
        # 已使用矩阵运算加速，max函数估计为O(logn)，总时间复杂度为O(nlogn)

    ## 中值滤波
    # temp = [Ret(CasterLevel,1); Ret(:,1); Ret(1,1)]
    # Ret(:,1) = medfilt1(temp,3)
    # Ret(:,1) = medfilt1(Ret(:,1),3)

    ## 模长归一化
    SquareSum = sum(Ret(:,1).^2)
    Ret(:,1) = Ret(:,1)/SquareSum*CasterLevel

    ## 目前问题：
    # 当交点消失时，此算法会搜索一个附近的较大值来取代通常交点